homeassistant/components/zha/config_flow.py

```diff
@@ -27,24 +45,177 @@
+        self._radio_type: RadioType | None = None
+        self._title: str | None = None
+        self._current_settings: zigpy.backups.NetworkBackup | None = None
+        self._backups: list[zigpy.backups.NetworkBackup] | None = None
+        self._chosen_backup: zigpy.backups.NetworkBackup | None = None
+
+        self.hass: HomeAssistant
+
+    @contextlib.asynccontextmanager
+    async def _connect_zigpy_app(self) -> ControllerApplication:
+        """Connect to the radio with the current config and then clean up."""
+        assert self._radio_type is not None
+
+        config = self.hass.data.get(DATA_ZHA, {}).get(DATA_ZHA_CONFIG, {})
+        app_config = config.get(CONF_ZIGPY, {}).copy()
+
+        database_path = config.get(
+            CONF_DATABASE,
+            self.hass.config.path(DEFAULT_DATABASE_NAME),
+        )
 
-    async def async_step_user(self, user_input=None):
-        """Handle a zha config flow start."""
-        if self._async_current_entries():
-            return self.async_abort(reason="single_instance_allowed")
+        # Don't create `zigbee.db` if it doesn't already exist
+        if not os.path.exists(database_path):
+            database_path = None
 
+        app_config[CONF_DATABASE] = database_path
+        app_config[CONF_DEVICE] = self._device_settings
+        app_config = self._radio_type.controller.SCHEMA(app_config)
+
+        app = await self._radio_type.controller.new(
+            app_config, auto_form=False, start_radio=False
+        )
+
+        try:
+            await app.connect()
+            yield app
+        finally:
+            await app.disconnect()
+
+    async def _restore_backup(
+        self, backup: zigpy.backups.NetworkBackup, **kwargs: dict[str, Any]
+    ) -> None:
+        """Restore the provided network backup, passing through kwargs."""
+        if self._current_settings is not None and self._current_settings.supersedes(
+            self._chosen_backup
+        ):
+            return
+
+        async with self._connect_zigpy_app() as app:
+            await app.backups.restore_backup(backup, **kwargs)
+
+    async def _detect_radio_type(self) -> bool:
+        """Probe all radio types on the current port."""
+        for radio in AUTOPROBE_RADIOS:
+            _LOGGER.debug("Attempting to probe radio type %s", radio)
+
+            dev_config = radio.controller.SCHEMA_DEVICE(
+                {CONF_DEVICE_PATH: self._device_path}
+            )
+            probe_result = await radio.controller.probe(dev_config)
+
+            if not probe_result:
+                continue
+
+            # Radio library probing can succeed and return new device settings
+            if isinstance(probe_result, dict):
+                dev_config = probe_result
+
+            self._radio_type = radio
+            self._device_settings = dev_config
+
+            return True
+
+        return False
+
+    async def _async_create_radio_entity(self):
```

Comment:
I recommend completing type annotations for the changed code in the module when working on the integration.